### DISCUSSION TITLE
Update from 7.0 to 8.2.4 SaveAsync fails due to object being IsBusy = true after Rule is hit with Fetch command/UOW

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#4096] [Questions] [answered]
Hello All,
We recently update from 7.0 to 8.2.4 and we encounter some issues while we try to save our business objects. What we noticed is that after we set our business objects properties and the a business rule kicks off the system would hang. We got a error stating that that we could not save due to object IsBusy = true. 
After further looking into it we notice that the business rules was hanging the application. Our business rule has a Command/UOW to fetch a list of ReadOnlyListBase objects from the database and notice that this call was causing the issue. We figured that maybe the Async method was not formulated correctly. So we made a simple method call within in the business rules. After further review its seems getting the results from the task is causing an issue. Can someone confirm if the code below should work or if we are doing something wrong. 

```c#
using Csla.Core;
using Csla.Rules;
using Csla;

namespace Pcdl.Biz.BusinessRules
{
    public class TestBusinessRule : BusinessRuleAsync
    {

        //Note the rule will not run if any of the input properties are empty or null
        //Make sure all methods are Async if you are using async methods. It will not work if you are using sync methods
        public TestBusinessRule(
                                            IPropertyInfo tParcelOldMBProperty,
                                            IPropertyInfo tParcelOldPGProperty,
                                            IPropertyInfo tParcelOldPCLProperty)
            : base(tParcelOldMBProperty)
        {
            //Note Order make a difference with InputProperties
            //PrimaryProperty = tParcelOldMBProperty;
            InputProperties.Add(tParcelOldMBProperty);
            InputProperties.Add(tParcelOldPGProperty);
            InputProperties.Add(tParcelOldPCLProperty);
        }

        protected override async Task ExecuteAsync(IRuleContext context)
        {
            string OldMBValue = (string)context.InputPropertyValues[InputProperties[0]];
            string OldPGValue = (string)context.InputPropertyValues[InputProperties[1]];
            string OldPCLValue = (string)context.InputPropertyValues[InputProperties[2]];

            if (!string.IsNullOrEmpty(OldMBValue) && !string.IsNullOrEmpty(OldPGValue) && !string.IsNullOrEmpty(OldPCLValue))
            {
                IDataPortal<ParcelNewList> dpPNLCommand = context.ApplicationContext.GetRequiredService<IDataPortal<ParcelNewList>>();
                //This failed I figured maybe the Async method was not set up correctly. Lets make it simple.
                //ParcelNewList objPNList = await dpPNLCommand.FetchAsync(OldMBValue, OldPGValue, OldPCLValue);

                ParcelNewList objPNList2 = null;

                //This fails. Everthing else is skipped and hangs. Set business object to IsBusy = true
                //objPNList2 = await dpPNLCommand.CreateAsync();

                //Created GetParcelNewListTask method to see if I can get the Task to work. 
                //This worked and returned the Task
                var objPNList = GetParcelNewListTask(dpPNLCommand, OldMBValue, OldPGValue, OldPCLValue);

                string test = "did it work";

                try
                {
                    //This fails. When I try to get the results from the task. Everthing else is skipped and hangs. Sets business object to IsBusy = true
                    objPNList2 = objPNList.Result;
                }
                catch (Exception ex)
                {
                    //No error is thrown. This is skiped and exit out 
                    string test1 = ex.Message;
                }

                string test2 = "did it work";

                if (OldMBValue != "0000" && OldPGValue != "0000" && OldPCLValue != "0000")
                {
                    foreach (ParcelNewInfo objPN in objPNList2)
                    {
                        //Do something here with the list
                    }
                }
            }
        }

        private Task<ParcelNewList> GetParcelNewListTask(IDataPortal<ParcelNewList> UOW, string MBSearch, string PGSearch, string PCLSearch)
        {
            //This works and returns the Task
            var objPNList = Task.FromResult(UOW.CreateAsync());
            return objPNList.Result;
        }
    }
}

```

**Comment by jhuerta3** [comment]
I noticed that my pages did not have a the latest. So Im updating everything now. CSLA to 8.2.5 and .NetCore to 8.0.7

**Comment by jhuerta3** [comment]
FYI - I updated all the Libraries to latest CSLA to 8.2.5 and .NetCore to 8.0.7 and I get the same results.

**Comment by kcabral817** [comment]
I have had no luck performing an external call like you've done in my business object async.  Particularly when using a factory class. I get the same results—a hang or immediate return with no call performed.

What I ended up doing was removing the Async from the name of the inherited class call and then doing this in a non-async method:

obj = Task.Run(async () => await portal.FetchAsync(criteria)
    .ConfigureAwait(false))
    .GetAwaiter()
    .GetResult();

I have never gotten async rules to work when calling a factory class to get data.

Have a nice day.

Kevin



________________________________

**Reply by jhuerta3** [reply]
Thank you @kcabral817. I will try that. Keep you posted.

**Reply by StefanOssendorf** [reply]
@kcabral817 What do you mean with `factory class` here? We are fetching new objects in rules without any problems.

**Comment by rockfordlhotka** [comment]
There are some capabilities in CSLA 8 to help make async rules easier to work with.

One thing is `BusinessRules.CheckRulesAsync`, which you should use instead of the old `CheckRules` method. The new one is something you can await, so it won't move forward until all async rules are complete.

This is particularly useful in, for example, a fetch operation method where you want to run the rules after loading the data:

```csharp
  [Fetch]
  private async Task FetchAsync(...)
  {
    // load the data into the object
    await BusinessRules.CheckRulesAsync();
  }
```

That'll ensure the async rules are finished before the object graph is returned to the client.

On the client side, the same thing is true. If your code calls `BusinessRules.CheckRules`, change it to the async version so you can await it.

Finally, on the client side, if you aren't manually saying to check rules, but allowing the rules engine to "do its thing" - the most common scenario in my experience, there are some things you can do.

First, `IsSaveable` won't be true if the object is busy, you can bind the save button's disabled state to `IsSaveable`. Recently I've learned that disabling buttons can be an accessibiity issue though, so keep that in mind.

Second, you can use the `WaitForIdle` method on the domain object:

```c#
  private async Task SaveButtonClicked()
  {
    await obj.WaitForIdle();
    await obj.SaveAndMergeAsync();
  }
```

The `WaitForIdle` method returns a Task that won't complete until the object is no longer busy.

**Reply by jhuerta3** [reply]
thank you @rockfordlhotka . I made some changes and it seems to be working but now my insert failed. Actually both approaches worked. Im going focus on the later approach. I will look further into why my insert is failing. It hits the Insert and enters the record but once it comes to the end of the method it hangs. To the next issue! I need to make sure Im following the sample you provided everywhere that might fix my issue.

**Reply by rockfordlhotka** [reply]
If it helps, the `WaitforIdle` has an optional timeout parameter. I think `CheckRulesAsync` does too.

The defaults might be a max TimeSpan, but it _should_ be possible to avoid hangs. If not, that's a bug.

**Comment by StefanOssendorf** [comment]
The business rule looks partially okay to me. I would _assume_ the `dpPNLCommand` is an actual command and should have an `ExecuteAsync()` and not `FetchAsync()`.
Where are the rules exeucted? Client or server? It's easy to deadlock your client (e.g. WPF/WinForms) when using `.Result`. So that must be avoided at all costs.

**Comment by kcabral817** [comment]
When I do this:

```csharp
private async Task GetBidHolidayInformationAsync(IRuleContext context)
{
   var portal1 = context.ApplicationContext.GetRequiredService<IDataPortal<ConstructionBidPlannedHolidayInfoList>>();
    BidStartDateEndDateCriteria criteria = new BidStartDateEndDateCriteria(bidKey, beginDate, endDate);
   var obj1 = await portal1.FetchAsync(criteria);
   totalHolidayCount = (short)obj1.Count();
}
```


It invokes this:
```csharp
    [Serializable]
    [Csla.Server.ObjectFactory(typeof(ConstructionBidPlannedHolidayFactory), "GetAllReadOnlyAsync")]
    public class ConstructionBidPlannedHolidayInfoList : ReadOnlyListBase<ConstructionBidPlannedHolidayInfoList, ConstructionBidPlannedHolidayInfo>
    {
        public ConstructionBidPlannedHolidayInfoList()
        {
            AllowNew = true;
            AllowRemove = true;
            AllowEdit = true;
            IsReadOnly = false;
        }
    }
}
```

Which calls this:
```csharp
 public async Task<object> GetAllReadOnlyAsync(int bidId)
 {
     var task1 = await _repository.GetAllAsync(bidId);

     var listing = _context.CreateInstanceDI<ConstructionBidPlannedHolidayInfoList>();

     SetIsReadOnly(listing, false);

     foreach (var item in task1)
     {
         var obj = _mapContainer.Map<ConstructionBidPlannedHoliday, ConstructionBidPlannedHolidayInfo>(item);

         obj.HolidayName = await GetHolidayAsync(obj.CompanyHolidayKey);
         MarkOld(obj);
         MarkAsChild(obj);
         listing.Add(obj);
     }

     SetIsReadOnly(listing, true);

     return listing;
 }
```

One of several things will happen:

  1.  var obj1 = await portal1.FetchAsync(criteria);   IMMEDIATE RETURN—No invocation of factory classes
  2. Return is null
  3. the fetch never returns and the cpu hovers around 2%​



________________________________

**Reply by StefanOssendorf** [reply]
Did you try to use a csla opreation annotaded with `[Fetch]` instead of the factory object? If not could you try that please? I'm not very familiar with the factory object approach.

**Comment by rockfordlhotka** [comment]
I'm guessing your `GetAllReadOnlyAsync` method isn't bypassing the rules as the child objects are loaded, so as the properties are set all the rules are being invoked - sync and async. That may be what you want.

If you have async rules, you need to wait for them to complete before exiting the method. `BusinessBase` has a `WaitForIdle` method that you can await so you know when the rules are done.

You could call `WaitForIdle` in each iteration of the foreach loop. Or you can get its Task and add it to a list of tasks, then wait for all those tasks after exiting the foreach loop.

**Comment by kcabral817** [comment]
But that class is a businessReadOnly, so there shouldn't be invocation of the rules.  Is that perspective correct?

Have a nice day.

Kevin



________________________________

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka, I made suggested changes but Im still having issues with the Insert process. I have two issues. The First is I am not able to retrieve the key value after the data is saved into the database. The second is my application hanges after the Business object Insert method ends. Can you verify my code and see if its correct.  The good thing is I dont get IsBusy = true any more after I added WaitForIdle  


`


	//UI Razor button save method 	

    private async Task<bool> SaveParcelChangeDataAsync()
    {            
            //NOTE: The code below is the new way to save the ParcelChangeEdit object
            //PageParcelChangeEdit = await PageParcelChangeEdit.SaveAsync();

            await PageParcelChangeEdit.WaitForIdle();
            await PageParcelChangeEdit.SaveAndMergeAsync();
       
          //other code here

       return result;

    }



	//Business Object Insert 

        [Insert]
        private async void InsertAsync([Inject] IParcelChangeDal dal)
        {
            using (BypassPropertyChecks)
            {
                var data = new ParcelChangeEntity
                {
			// Save the properties
                };
                var result = await dal.InsertAsync(data);
		//The result.tParcelChangeID is always zerro 
                tParcelChangeID = result.tParcelChangeID;
            }
	//Hangs after this method ends.
        }


	//Data Access Layer Insert

        public Task<ParcelChangeEntity> InsertAsync(ParcelChangeEntity data)
        {
            lock (_context)
            {                          
                //_context.ParcelChanges.Add(data);
                //_context.SaveChanges();
		
		//The record is saved into the database but my Key is Zero on the object. 
		//Although the Key exist in the database. 
                _context.ParcelChanges.AddAsync(data);
                _context.SaveChangesAsync();
            }
		//data.tParcelChangeID is zero. The Key value from the database is not in the object.
            return Task.FromResult(data);
        }


`

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka ,
I perform a test and re-added the orginal lines to save the record and the Key value is in the object. Its the Async calls that possible dont work. 

```
	//Data Access Layer Insert
        public Task<ParcelChangeEntity> InsertAsync(ParcelChangeEntity data)
        {
            lock (_context)
            {
                     
		//This works!      
                _context.ParcelChanges.Add(data);
                _context.SaveChanges();
		
		//The record is saved into the database but my Key is Zero on the object. 
		//Although the Key exist in the database. 
                //_context.ParcelChanges.AddAsync(data);
                //_context.SaveChangesAsync();
            }
		//data.tParcelChangeID is zero. The Key value from the database is not in the object.
            return Task.FromResult(data);
        }
```

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka ,

The following below works fine. I am able to save to the database and the code continues to process in the UI but I had to remove all the Async calls and syntax. Maybe something is wrong with my syntax.

`
     //UI Razor button save method 	
    //Removed Async and this works the call will continue. 
    private bool SaveParcelChangeData()
    {
        bool result = false;

	    //This works!
            PageParcelChangeEdit = PageParcelChangeEdit.Save();

            //NOTE: The code below is the new way to save the ParcelChangeEdit object
            //PageParcelChangeEdit = await PageParcelChangeEdit.SaveAsync();

            // await PageParcelChangeEdit.WaitForIdle();
            // await PageParcelChangeEdit.SaveAndMergeAsync();
            // await PageParcelChangeEdit.WaitForIdle();

            return result;
    }


	//Business Object Insert 
	//Removed Async and this works the call will continue. 
        [Insert]
        private void Insert([Inject] IParcelChangeDal dal)
        {

            ClaimsPrincipal principal = this.ApplicationContext.Principal;
            string? _CreatedByID = principal.Claims.FirstOrDefault(c => c.Type == "EmployeeNumber")?.Value;

            using (BypassPropertyChecks)
            {
                var data = new ParcelChangeEntity
                {
			//Set Properties
                };
                var result = dal.Insert(data);
                tParcelChangeID = result.tParcelChangeID;
            }
        }



	//Data Access Layer Insert
	//Removed Async and this works the call will continue. 
        public ParcelChangeEntity Insert(ParcelChangeEntity data)
        {
            lock (_context)
            {
		//This works!                            
                _context.ParcelChanges.Add(data);
                _context.SaveChanges();

                //This does not work. The ID is always 0
                //_context.ParcelChanges.AddAsync(data);
                //_context.SaveChangesAsync();
            }
            return data;
        }


`

**Reply by rockfordlhotka** [reply]
Here's some code I just created to test the overall flow. Not using EF, I'm not an EF expert, so can't help with getting the id back from EF. But all the CSLA stuff works here.

```c#
using Csla;
using Csla.Configuration;
using Microsoft.Extensions.DependencyInjection;
using MyCode;

var services = new ServiceCollection();
services.AddCsla();
services.AddTransient<ParcelDal>();
var serviceProvider = services.BuildServiceProvider();

var portal = serviceProvider.GetRequiredService<IDataPortal<ParcelEdit>>();
var edit = await portal.CreateAsync();
Console.WriteLine(edit);
edit.Name = "New Parcel";
Console.WriteLine(edit);
await edit.SaveAndMergeAsync();
Console.WriteLine(edit);

namespace MyCode
{
    public class ParcelEdit : BusinessBase<ParcelEdit>
    {
        public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
        public int Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }

        public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
        public string Name
        {
            get => GetProperty(NameProperty);
            set => SetProperty(NameProperty, value);
        }

        public override string ToString()
        {
            return $"Id {Id}, Name '{Name}', IsNew {IsNew}, IsDirty {IsDirty}";
        }

        [Create]
        private async Task CreateAsync([Inject] ParcelDal dal)
        {
            var entity = await dal.CreateAsync();
            using (BypassPropertyChecks)
            {
                Id = entity.Id;
                Name = entity.Name;
            }
            BusinessRules.CheckRules();
        }

        [Fetch]
        private async Task FetchAsync([Inject]ParcelDal dal)
        {
            var entity = await dal.FetchAsync();
            using (BypassPropertyChecks)
            {
                Id = entity.Id;
                Name = entity.Name;
            }
        }

        [Insert]
        [Update]
        private async Task UpsertAsync([Inject]ParcelDal dal)
        {
            var entity = new ParcelEntity { Id = Id, Name = Name };
            if (IsNew)
                entity = await dal.InsertAsync(entity);
            else
                entity = await dal.UpdateAsync(entity);
            using (BypassPropertyChecks)
            {
                Id = entity.Id;
            }
        }
    }

    public class ParcelDal
    {
        private static int _id = 0;

        public async Task<ParcelEntity> CreateAsync()
        {
            return new ParcelEntity();
        }

        public async Task<ParcelEntity> FetchAsync()
        {
            return new ParcelEntity { Id = ++_id, Name = "Parcel" };
        }

        public async Task<ParcelEntity> InsertAsync(ParcelEntity entity)
        {
            // Insert entity
            entity.Id = ++_id;
            return entity;
        }

        public async Task<ParcelEntity> UpdateAsync(ParcelEntity entity)
        {
            // Update entity
            return entity;
        }
    }

    public class ParcelEntity
    {
        public int Id { get; set; }
        public string Name { get; set; } = string.Empty;
    }
}
```

**Reply by rockfordlhotka** [reply]
Though shame on me, because I didn't use `CheckRulesAsync` like I should:

```c#
            await BusinessRules.CheckRulesAsync();
```

**Comment by jhuerta3** [comment, marked as answer]
Hello @rockfordlhotka and @kcabral817 ,

I finally got my code working. This is what I had to do to get it to work and what I found. 
1. - If im loading a business object properties with data that will execute a business Async rule then the method loading the data on the object must also be Async so that I can call object.WaitForIdle() and it work;
1.1 - What I notice is that once the first property is loaded with data that kicks off a business rule the object.IsBusy = true. 
1.2 – After loading all the data then call object.WaitForIdle() to set the object.IsBusy = false;

2. The next thing is I had to set the insert methos and save methods to Sync and not Async. 
2.1 – If I didn’t do this it would hang on the business object Insert for my case. 

3.  Also I had a business rule that was looking at an extra property on the object that was not being used in the business rules.  That extra property was also causing the object.IsBusy = true and object.WaitForIdle(); would not work. 
3.1 - Once I removed the extra property from the business rule was able to save. 
3.1 - Im guesing since the property was not being used in the rule it would not switch from object.IsBusy = true to object.IsBusy = false. 

Below is my code. I will research more moving forward. If someone can use these notes good.  Thank you for you help. 


`

    //Blazor Razor UI method I had to make it async 
    private async Task Save()
    {

            PageParcelNewEdit = await LoadParcelNewDataAsync();

            if (PageParcelNewEdit != null)
                await PageParcelNewEdit.WaitForIdle();

            isSaved = await SaveParcelDataAsync();
            StateHasChanged();

    }



    //Blazor Razor UI method I had to make it async so that await objParcelNE.WaitForIdle(); would work.
    private async Task<ParcelNewEdit> LoadParcelNewDataAsync(ParcelNewEdit objParcelNE)
    {

	//These properties kick off business rules which is a Async Rule
            objParcelNE.tParcelNewMB = NewMB;
            objParcelNE.tParcelNewPG = NewPG;
            objParcelNE.tParcelNewPCL = NewPCL;
            //Removed this property since it was not used in the business rules and setting the object to objParcelNE.IsBusy = true
            //objParcelNE.tParcelChangeID = PageParcelChangeEdit.tParcelChangeID;

	//I had to add this to check if its IsBusy to mark it false
        if (objParcelNE.IsBusy)
        {
            await objParcelNE.WaitForIdle();
        }

        return objParcelNE;
    }


    private async Task<bool> SaveParcelDataAsync()
    {
        bool result = false;

        if (SaveNewParcelData())
        {
            ClearBrokenRules();
            ErrorMessagesList.Add("Parcel information has been saved.");
            ErrorMessageType = "info";
            return result = true;
        }
        else
            return result;
    }



    private bool SaveNewParcelData()
    {
        bool result = false;

        if (PageParcelNewEdit.IsValid)
        {
            try
            {
                PageParcelNewEdit = GetNewParcelAsync().Result;

                //I had to use this code since the Async calls did not work. It would hang. (For me)
                PageParcelNewEdit = PageParcelNewEdit.Save();

                //The code below does not work.
                // await PageParcelNewEdit.WaitForIdle();
                // await PageParcelNewEdit.SaveAndMergeAsync();

             }
         catch (Exception ex)
         {
             ErrorMessagesList.Add(ex.Message);
             ErrorMessageType = "Error";
         }
      }

      if (PageParcelNewEdit.BrokenRulesCollection.Count > 0)
      {
          CheckBrokenRules(PageParcelNewEdit.BrokenRulesCollection);
          return result;
      }

  }



	//Business Object - I had to make this Sync for it to work
	//It would hang if I did not
        [Insert]
        private void Insert([Inject] IParcelNewDal dal)
        {
            BusinessRules.CheckRules();

            using (BypassPropertyChecks)
            {
                var data = new ParcelNewEntity
                {
		   //Load Properties			
                };

                var result = dal.Insert(data);
                tParcelNewID = result.tParcelNewID;
            }
        }


	//Data Access object - I had to make this Sync for it to work	
        public ParcelNewEntity Insert(ParcelNewEntity data)
        {
            lock (_context)
            {
                //This would set my Key value that was set from database.    
                _context.ParcelNews.Add(data);
                _context.SaveChanges();

		//The record is saved into the database but my Key is always Zero on the object. 
		//Although the Key exist in the database. 
                //_context.ParcelNews.AddAsync(data);
                //_context.SaveChangesAsync();
            }
            return data;
        }


`

Thank you.

**Reply by StefanOssendorf** [reply]
Hi. There are some things I don't understand or can be solved in a more straight forward way.
```csharp
//I had to add this to check if its IsBusy to mark it false
if (objParcelNE.IsBusy)
{
    await objParcelNE.WaitForIdle();
}
```
can be directly written as
```csharp
await objParcelNE.WaitForIdle();
```
The `WaitForIdle` method checks if the object is still busy. If not it's a no-op method and immediatly returns.

This `PageParcelNewEdit = GetNewParcelAsync().Result;` should be properly awaited if it's an async method.

```csharp
[Insert]
    private void Insert([Inject] IParcelNewDal dal)
    {
        BusinessRules.CheckRules();
```
That last line kicks off all async rules without waiting for them to complete. You should change that to `await BusinessRules.CheckRulesAsync();` to make sure all rules are finished before starting the save preparations.

```csharp
public ParcelNewEntity Insert(ParcelNewEntity data)
    {
        lock (_context)
        {
            //This would set my Key value that was set from database.    
            _context.ParcelNews.Add(data);
            _context.SaveChanges();

	//The record is saved into the database but my Key is always Zero on the object. 
	//Although the Key exist in the database. 
            //_context.ParcelNews.AddAsync(data);
            //_context.SaveChangesAsync();
        }
        return data;
    }
```
I don't know why you lock on the db context but there's probably a good reason for that. This works in async (we are using it with a WPF client).
1. Don't use `AddAsync()` with EF core. Just use `Add()`. As long as you don't use the `SequenceHiLo` stuff from EF you don't need `AddAsync()` (c.f. [EF Core doc](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.addasync?view=efcore-8.0#microsoft-entityframeworkcore-dbcontext-addasync(system-object-system-threading-cancellationtoken) ))
2. In this code the `await` is missing before `SaveChangesAsync` so maybe that's the issue why you didn't get the updated ID value?

The "problem" with async is you have to go it all the way to your top most method.

**Comment by jhuerta3** [comment]
I forgot to add that I also added await BusinessRules.CheckRulesAsync(); to all my Async Fetch methods.
